File: set2q10.m
clear all
close all
img=imread('cameraman.tif'); 
[M N]=size(img);  
p=2*M; 
q=2*N; 
%FT of image centered 
F=fftshift(fft2(img,p,q)); 
%centre of the frequency rectangle 
cX=p/2; 
cY=q/2; 
%distance Metrix 
D=zeros(p,q); 
for i=1:p 
    for j=1:q 
        D(i,j)=sqrt((i-cX)^2+(j-cY)^2); 
    end
end

%ideal Low Pass Filter
D0=50; 
hid=(D<=D0); 
Ideal_Img=hid.*F; 
Ideal_Img=fftshift(Ideal_Img); 
Ideal_Img=real(ifft2(Ideal_Img)); 
Ideal_Img=Ideal_Img(1:M,1:N); 
subplot(2,2,1); 
imshow(img),title('Original Image'); 
subplot(2,2,2); imshow(uint8(Ideal_Img)),title('Ideal LPF output'); 

%Butterworth Filter 
n=5; %order
h_bu=zeros(p,q); 
for i=1:p 
    for j=1:q 
        h_bu(i,j)=1/(1+(D(i,j)/D0)^(2*n)); 
    end
end
btrWorth=h_bu.*F; 
btrWorth=fftshift(btrWorth); 
btrWorth=real(ifft2(btrWorth)); 
btrWorth=btrWorth(1:M,1:N); 
subplot(2,2,3); imshow(uint8(btrWorth)); title('Butterworth LPF Output');

%Guassian Low Pass Filter
G=zeros(p,q);
for i=1:p
    for j=i:q
        G(i,j)=exp(-D(i,j)^2/(2*D0^2));
    end
end
gsflt=G.*F;
gsflt=fftshift(gsflt);
gsflt=real(ifft2(gsflt));
gsflt=gsflt(1:M,1:N);
subplot(2,2,4);
imshow(uint8(gsflt));
title('Gaussian LPF Output');

File: set2q10.py
import numpy as np
import matplotlib.pyplot as plt
import imageio

im=imageio.imread('cameraman.tif'); 
M,N=im.shape;  
p=2*M; 
q=2*N; 
pd1=np.int(M/2)
pd2=np.int(N/2)
img=np.pad(im,[pd1,pd2],'constant')
#FT of image centered 
F=np.fft.fftshift(np.fft.fft2(img)); 
#centre of the frequency rectangle 
cX=p/2; 
cY=q/2; 
#distance Metrix 
D=np.zeros((p,q)); 
for i in range(p):
    for j in range(q):
        D[i,j]=np.sqrt((i-cX)**2+(j-cY)**2)

#ideal Low Pass Filter 
D0=50; 
hid=(D<=D0); 
Ideal_Img=np.multiply(hid,F); 
Ideal_Img=np.fft.fftshift(Ideal_Img); 
Ideal_Img=np.real(np.fft.ifft2(Ideal_Img)); 
Ideal_Img=Ideal_Img[int(M/2):int(3*M/2),int(N/2):int(3*N/2)]; 
plt.subplot(2,2,1); 
plt.imshow(im,cmap='gray'),plt.title('Original Image'); 
plt.axis('off')
plt.subplot(2,2,2); plt.imshow(Ideal_Img,cmap='gray'),plt.title('Ideal LPF output'); 
plt.axis('off')

#Butterworth Filter 
n=5; #order
h_bu=np.zeros((p,q)); 
for i in range(p):
    for j in range(q):
        h_bu[i,j]=1/(1+(D[i,j]/D0)**(2*n))
btrWorth=np.multiply(h_bu,F); 
btrWorth=np.fft.fftshift(btrWorth); 
btrWorth=np.real(np.fft.ifft2(btrWorth)); 
btrWorth=btrWorth[int(M/2):int(3*M/2),int(N/2):int(3*N/2)]
#btrWorth=btrWorth[1:M,1:N]; 
plt.subplot(2,2,3); plt.imshow(btrWorth,cmap='gray'); plt.title('Butterworth LPF Output');
plt.axis('off')

#Guassian Low Pass Filter
G=np.zeros((p,q));
for i in range(p):
    for j in range(q):
        G[i,j]=np.exp(-D[i,j]**2/(2*D0**2));
gsflt=np.multiply(G,F);
gsflt=np.fft.fftshift(gsflt);
gsflt=np.real(np.fft.ifft2(gsflt));
gsflt=gsflt[int(M/2):int(3*M/2),int(N/2):int(3*N/2)];
plt.subplot(2,2,4);
plt.imshow(gsflt,cmap='gray');
plt.title('Gaussian LPF Output');
plt.axis('off')
plt.show()